### PR TITLE
Use recent-rejects, orphans, and recently-replaced txn for compact-block-reconstruction

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This takes the rather obvious step of keeping a circular buffer of CTransactionRefs which is populated with orphan transactions, transactions replaced out of mempool, and rejected transactions and uses them to populate compact blocks.

It defaults to 100 transactions in size, the same as the orphan map, which, in the worst case, would be about 40MB+memory overhead (this would be 100 400K-weight witness-only-ish transactions replaced, same as the orphan map).

It limits rejected transactions to 100KB in memory usage before storing.

In previous tests, @gmaxwell had indicated that if we had an infinite-sized buffer here we could save a ton of compact-block-round-trips, so in non-adversarial conditions even a small buffer should help at least sometimes.

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Most of the diff here is changing the signature of AcceptToMemoryPool trivially....

**Comment by TheBlueMatt** [comment]
To clarify: only rejected transactions are limited based on memory-usage (to 100KB), everything else relies on the standardness limits - 400K weight, which would be 400KB of witness data, max.

**Comment by gmaxwell** [comment]
Concept ACK. nice implementation.

You should probably consider adding an extra count "% from memory (% hits against extra)," to the successful reconstruction log entry.

The count limit is unfortunate, merely 100 entries but 40MB max?  you could easily keep count of the worst case memory usage (ignoring the deduplication) and then just target 4MB-- lots less memory and a higher hitrate no doubt.

**Comment by TheBlueMatt** [comment]
Indeed, doing smarter limiting would be nice, but this should do for a bare-bones help-the-default-case win for 0.14 (if it turns out to do enough and the fast progress we've been making on review continues through the end of the week). If we prefer it's not really any additional review burden to do a memory-based limit across the entire list.

**Review by instagibbs — APPROVED** [review]
very nice, and plenty of room for improvement on top later

utACK, running now with an extra commit on top to see efficacy: https://github.com/instagibbs/bitcoin/commit/fcb8ace8245435289018a8171463f4743ea664a5

**Comment by gmaxwell** [comment]
It's working for me. I added logging for transactions that come from the extrapool

> +    LogPrint("cmpctblock", "Successfully reconstructed block %s with %lu txn prefilled, %lu txn from memory (%lu extra hits), and %lu txn requested\n", hash.ToString(), prefilled_count, mempool_count, extra_count, vtx_missing.size());

Please add that, I know extra can't be counted completely precisely in the presence of collisions but it's very useful information.

Some stats: 

$ grep essfu ~/.bitcoin/debug.log | tail -n 72  | awk '{aa=int(substr($15,2))>0; bb=$19>0; if (aa && !bb) {saved+=1} else if (!aa && !bb) {noextra+=1} else {roundtrip+=1} } END {print saved/NR " " noextra/NR " " roundtrip/NR}' 
0.180556 0.527778 0.291667

So the extra increased no-roundtrip from 52% to 68%... in the last 72 blocks here. Though I see there are at least a couple blocks where extra was just too small and would have been successful if it were larger.  I think thats a big enough improvement to justify doing it, but the size should be increased eventually (presumably via better memory management).

**Comment by TheBlueMatt** [comment]
Renamed vOrphanAndRecentlyRemovedTxn -> vExtraTxnForCompact, no other changes

File: src/net_processing.cpp [hunk 1]
```diff
@@ -1741,7 +1741,10 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                 // See https://github.com/bitcoin/bitcoin/issues/8279 for details.
                 assert(recentRejects);
                 recentRejects->insert(tx.GetHash());
-            }
+                if (RecursiveDynamicUsage(*ptx) < 100000)
```
Comment (morcos) [inline src/net_processing.cpp]: A quick examination shows 80-90% of transactions not accepted to the mempool are due to rate limited free transaction.   What do you think about also adding to this check and the one below `&& state.GetRejectCode() != REJECT_INSUFFICIENTFEE`

You could be throwing away some transactions that would help blocks reconstruct, but I get close to 100 "not accepted" transactions an hour and it seems like I'd rather have 5 hours of non-free transactions rather than 1 hour of all.

Comment (gmaxwell) [inline src/net_processing.cpp]: It might be more interesting to distinguish REJECT_INSUFFICIENTFEE from rejection because it didn't even meet the static minimum minfee.

Comment (TheBlueMatt) [inline src/net_processing.cpp]: I was trying to avoid any complication here, but if you think something simple is worth it I'm ok with adding something.

Comment (morcos) [inline src/net_processing.cpp]: Yep understood, my guess was that this was simple and a decent improvement but @gmaxwell has a point in that failed replacements could also be REJECT_INSUFFICIENTFEE and those might not be things you want to throw away.  So up to you if you can simply disambiguate.

Comment (gmaxwell) [inline src/net_processing.cpp]: I propose leaving it alone for now, in the long run we'll want some kind of smarter queue.

**Comment by sdaftuar** [comment]
> So the extra increased no-roundtrip from 52% to 68%... in the last 72 blocks here. 

I did a test on data from the second half of November (11/16-11/30) to see what the improved reconstruction rate would have been on the data my logging node saw, and I saw a gain as well, though more modest than the one you observed -- 923 (out of 2137) blocks in that window reconstructed with no round trip using the old code, and 984 blocks reconstructed with no round-trip after this PR.

**Comment by gmaxwell** [comment]
My last 72 blocks: 0.125 0.625 0.25  (so .62 to .75)

**Comment by instagibbs** [comment]
(piling on) with a 1000 transaction buffer: 0.0833333 0.597222 0.319444 (~8% absolute improvement)

File: src/blockencodings.cpp [hunk 2]
```diff
@@ -47,7 +47,7 @@ uint64_t CBlockHeaderAndShortTxIDs::GetShortID(const uint256& txhash) const {
 
 
 
-ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& cmpctblock) {
+ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& cmpctblock, std::vector<std::pair<uint256, CTransactionRef>>& extra_txn) {
```
Comment (theuni) [inline src/blockencodings.cpp]: const? Or was this intended to remove the matches?

**Comment by morcos** [comment]
1000 tx buffer I got
0.19 0.65 0.16

Definitely worth merging I say..

ACK 863edb4 , but please also add the extra printing

**Comment by instagibbs** [comment]
tACK https://github.com/bitcoin/bitcoin/pull/9499/commits/863edb45b9841c3058e883015c7576e6f69e3cc6 (agree with additional printing)

**Comment by TheBlueMatt** [comment]
Made the second parameter to InitData const and added the requested debug print.

File: src/blockencodings.cpp [hunk 3]
```diff
@@ -130,6 +131,37 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
...
+                    extra_count--;
+                }
+            }
+        }
+        // Though ideally we'd continue scanning for the two-txn-match-shortid case,
+        // the performance win of an early exit here is too good to pass up and worth
+        // the extra risk.
+        if (mempool_count == shorttxids.size())
+            break;
+    }
```
Comment (luke-jr) [inline src/blockencodings.cpp]: A lot of code duplication here that would probably be better shared, but maybe not critical for 0.14.

**Comment by luke-jr** [comment]
I wonder if it might be better to simply cache whether transactions we discard are valid or invalid, rather than the full transaction data. Then hypothetically we can judge the entire block as valid/invalid quickly, even though we still have more data to fetch. But this idea sounds like a lot of work...

File: src/blockencodings.cpp (hunk 3 above)
Comment (TheBlueMatt) [inline src/blockencodings.cpp]: Yea, I thought about it, but to avoid having two vectors I decided against it. Will fix in a follow-up pr for post-0.14.

**Comment by TheBlueMatt** [comment]
@luke-jr we tried that before...and would have broken consensus had it been shipped, so, yes, that is "a lot of work".

**Comment by gmaxwell** [comment]
ACK.

File: src/blockencodings.cpp [hunk 4]
```diff
@@ -130,6 +131,35 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
         if (mempool_count == shorttxids.size())
             break;
     }
+    }
```
Comment (sipa) [inline src/blockencodings.cpp]: Maybe add a comment explaining that if a short txid appears in both the mempool and in the extra_txn vector, we still assume the mempool one is the correct one, and don't treat it as a collision. As the mempool has much better chance for matching with actual blocks, this is justified.

File: src/net_processing.cpp [hunk 5]
```diff
@@ -1741,7 +1741,10 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
...
+                    AddToCompactExtraTransactions(ptx);
+            } else if (tx.HasWitness() && RecursiveDynamicUsage(*ptx) < 100000)
```
Comment (sipa) [inline src/net_processing.cpp]: This branch looks like it will never be taken, as every time the condition holds, the previous branch should have matches as well.

Comment (sipa) [inline src/net_processing.cpp]: Nevermind. I'm confused by the the braces. Can you instead please add braces around all the indented sections?

File: src/blockencodings.cpp (hunk 4 above)
Comment (TheBlueMatt) [inline src/blockencodings.cpp]: This isnt true? This is only true if the witness hash (and thus, the txn) are identical.

Comment (TheBlueMatt) [inline src/blockencodings.cpp]: In fact, there's already a comment indicating this.

File: src/blockencodings.cpp [hunk 6]
```diff
@@ -130,6 +131,35 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
...
+    for (size_t i = 0; i < extra_txn.size(); i++) {
+        uint64_t shortid = cmpctblock.GetShortID(extra_txn[i].first);
+        std::unordered_map<uint64_t, uint16_t>::iterator idit = shorttxids.find(shortid);
+        if (idit != shorttxids.end()) {
+            if (!have_txn[idit->second]) {
+                txn_available[idit->second] = extra_txn[i].second;
+                have_txn[idit->second]  = true;
+                mempool_count++;
+            } else {
+                // If we find two mempool txn that match the short id, just request it.
```
Comment (TheBlueMatt) [inline src/blockencodings.cpp]: OK, added.

Comment (TheBlueMatt) [inline src/blockencodings.cpp]: Note that the comment had already been mostly adapted, with an extra few lines added at the end to clarify that if there is a conflict between extra and mempool, it is ignored only if the transactions are identical.

File: src/net_processing.cpp (hunk 5 above)
Comment (TheBlueMatt) [inline src/net_processing.cpp]: OK, braces added.

File: src/blockencodings.cpp (hunk 4 above)
Comment (sipa) [inline src/blockencodings.cpp]: Seems I was misreading the code, and interpreting the early-exit `mempool_count == shorttxids.size()` as the way the loop always successfully ends.

File: src/net_processing.cpp [hunk 7]
```diff
@@ -1723,7 +1741,10 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
...
```
Comment (luke-jr) [inline src/net_processing.cpp]: Shouldn't we at least check that it's *valid*?

Comment (luke-jr) [inline src/net_processing.cpp]: What stops someone from flooding peers with the same transaction triggering this case and making the extra-pool useless?

**Review by luke-jr — COMMENTED** [review]
utACK c73554042886fb63fb48edf29cf827951edde341 edded808fc4eee94c178e1779b90d1c87a08c23a 1531652e02d856c1b8c3ba6f6e51e1983ac0540c 93380c5247526e2248248a7d539233ec48d11bdd 7f8c8cab1e537809848088d3bfaa13ecca7ac8cc b55b4163462756cbb2341294b839a881333d8f1e fac4c78028d132ee7becc1dc54a0c881ae1a0287

Not so sure about 863edb45b9841c3058e883015c7576e6f69e3cc6

Disclaimer: Reviewed across 3 sittings

File: src/net_processing.cpp [hunk 8]
```diff
@@ -1723,6 +1741,11 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                 // See https://github.com/bitcoin/bitcoin/issues/8279 for details.
                 assert(recentRejects);
                 recentRejects->insert(tx.GetHash());
+                if (RecursiveDynamicUsage(*ptx) < 100000) {
+                    AddToCompactExtraTransactions(ptx);
+                }
+            } else if (tx.HasWitness() && RecursiveDynamicUsage(*ptx) < 100000) {
```
Comment (luke-jr) [inline src/net_processing.cpp]: Shouldn't we at least check that it's *valid*?

Comment (luke-jr) [inline src/net_processing.cpp]: What stops someone from flooding peers with the same transaction triggering this case and making the extra-pool useless?

Comment (TheBlueMatt) [inline src/net_processing.cpp]: is valid how? The CValidationState was likely not ever set if there were mempool conflicts (which is exactly the case we want here)

Comment (TheBlueMatt) [inline src/net_processing.cpp]: There isnt actually much you can do to prevent this - the goal is to capture some number of conflicting transactions and keep them around, and without some complicated rate-limiting logic you're not gonna win that fight.

Comment (luke-jr) [inline src/net_processing.cpp]: There are a few cases of invalid transactions we could detect; eg, script fails.

Comment (TheBlueMatt) [inline src/net_processing.cpp]: Yes, but we set state.Invalid for txn-mempool-conflict, and I super dont want a huge list of which cases of invalid we care about and which we dont here.

Comment (luke-jr) [inline src/net_processing.cpp]: Fair enough. I guess worst case it simply behaves the same as if this feature was never added.

File: src/net_processing.cpp [hunk 9]
```diff
@@ -59,6 +59,9 @@ map<uint256, COrphanTx> mapOrphanTransactions GUARDED_BY(cs_main);
 map<COutPoint, set<map<uint256, COrphanTx>::iterator, IteratorComparator>> mapOrphanTransactionsByPrev GUARDED_BY(cs_main);
 void EraseOrphansFor(NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
+static size_t vExtraTxnForCompactIt = 0;
```
Comment (TheBlueMatt) [inline src/net_processing.cpp]: Leaving it outside makes it clear that you cannot simply insert into ExtraTxn, you need to do something special (and, technically, avoids a hidden atomic operation).

File: src/net_processing.cpp (hunk 8 above)
Comment (TheBlueMatt) [inline src/net_processing.cpp]: Yea, this is no worse than before, and in many cases is a big win...its also written to be an easy review and merge, with the opportunity of cleaning it up and making it more clever in the future.

File: src/net_processing.cpp [hunk 10]
```diff
@@ -586,6 +589,17 @@ void UnregisterNodeSignals(CNodeSignals& nodeSignals)
 // mapOrphanTransactions
 //
 
+void AddToCompactExtraTransactions(const CTransactionRef& tx)
+{
```
Comment (TheBlueMatt) [inline src/net_processing.cpp]: Done

File: src/net_processing.cpp (hunk 8 above)
Comment (luke-jr) [inline src/net_processing.cpp]: Hmm, that sounds like something we probably shouldn't be doing (although off-topic for this PR). How about using the ban score?

**Comment by luke-jr** [comment]
utACK 863edb45b9841c3058e883015c7576e6f69e3cc6 too

File: src/net_processing.cpp (hunk 8 above)
Comment (TheBlueMatt) [inline src/net_processing.cpp]: Lets revisit this after we fix DoS banning and have a more reasonable API to access :).

File: src/net_processing.cpp (hunk 10 above)
Comment (TheBlueMatt) [inline src/net_processing.cpp]: Backed this out - some of our orphan map test stuff doesn't take cs_main, so it would fail in test_bitcoin (though all of the actual-bitcoind runs succeeded). This should be as a part of a separate PR to add cs_main and AssertLockHelds to all of the orphan handling functions.

**Comment by instagibbs** [comment]
re-ACK https://github.com/bitcoin/bitcoin/pull/9499/commits/c5945804ca7e5edd9fce8c6de9b1f0ef775f9e79

File: src/blockencodings.cpp [hunk 11]
```diff
@@ -130,6 +131,38 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
...
+                    txn_available[idit->second].reset();
+                    mempool_count--;
+                    extra_count--;
+                }
+            }
+        }
+        // Though ideally we'd continue scanning for the two-txn-match-shortid case,
+        // the performance win of an early exit here is too good to pass up and worth
+        // the extra risk.
+        if (mempool_count == shorttxids.size())
```
Comment (jnewbery) [inline src/blockencodings.cpp]: I don't think this early exit is worth it. The only way we could hit this is if we didn't find all of the shortid transactions in the mempool *and* we found the remaining shortid transactions in vExtraTxnForCompact. If that's the case, what do we gain/lose from early exit:

- gain: we won't have to scan through the < 100 transactions in vExtraTxnForCompact. That's an almost negligable win.
- lose: if there is a second transaction in vExtraTxnForCompact with the same shortid, we run the risk of a  READ_STATUS_FAILED in FillBlock.

File: src/blockencodings.cpp [hunk 12]
```diff
@@ -130,6 +131,38 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
...
+                // If we find two mempool/extra txn that match the short id, just
+                // request it.
+                // This should be rare enough that the extra bandwidth doesn't matter,
+                // but eating a round-trip due to FillBlock failure would be annoying
+                // Note that we dont want duplication between extra_txn and mempool to
+                // trigger this case, so we compare witness hashes first
+                if (txn_available[idit->second] &&
+                        txn_available[idit->second]->GetWitnessHash() != extra_txn[i].second->GetWitnessHash()) {
+                    txn_available[idit->second].reset();
+                    mempool_count--;
```
Comment (jnewbery) [inline src/blockencodings.cpp]: These counts are wrong if there's a triple shortid collision. Same for the mempool count above.

If there's a shortid collision between a transaction in the mempool and in vExtraTxnForCompact, I believe extra_count will underflow.

**Review by jnewbery — COMMENTED** [review]
Apologies for the post-merge review!

File: src/blockencodings.cpp (hunk 12 above)
Comment (TheBlueMatt) [inline src/blockencodings.cpp]: Indeed, looks like the counting will provide spurious results sometimes...want to open a new pr to fix it?

File: src/blockencodings.cpp (hunk 11 above)
Comment (TheBlueMatt) [inline src/blockencodings.cpp]: Looping through the GetShortID calls is nontrivial for large-ish mempools, and collisions should be ~1 in a million, so I think its still probably worth it.

Comment (TheBlueMatt) [inline src/blockencodings.cpp]: Plus the intention would be to significantly increase the size of the extra pool in 0.15, assuming smarter memory management.

Comment (jnewbery) [inline src/blockencodings.cpp]: Understood. When the extra pool size is 100 transactions this isn't worth it, but if the extra pool is made much larger in 0.15, it makes sense.

File: src/blockencodings.cpp (hunk 12 above)
Comment (jnewbery) [inline src/blockencodings.cpp]: Sure. I'll take a look tomorrow.
